langpuppet/gapi.go

```diff
@@ -0,0 +1,327 @@
+
+	dataLang := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(lang.Name+": "+format, v...)
+		},
+	}
+
+	dataPuppet := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var graphErr error
+
+	obj.graphFlagMutex.Lock()
+	if obj.langGraphReady {
+		obj.langGraphReady = false
+		obj.graphFlagMutex.Unlock()
+		obj.currentLangGraph, graphErr = obj.langGAPI.Graph()
+		if graphErr != nil {
+			return nil, graphErr
+		}
+	} else {
+		obj.graphFlagMutex.Unlock()
+	}
+
+	obj.graphFlagMutex.Lock()
+	if obj.puppetGraphReady {
+		obj.puppetGraphReady = false
+		obj.graphFlagMutex.Unlock()
+		obj.currentPuppetGraph, graphErr = obj.puppetGAPI.Graph()
+		if graphErr != nil {
+			return nil, graphErr
+		}
+	} else {
+		obj.graphFlagMutex.Unlock()
+	}
+
+	g, err := mergeGraphs(obj.currentLangGraph, obj.currentPuppetGraph)
+
+	if obj.data.Debug {
+		obj.currentLangGraph.Logf(func(format string, v ...interface{}) {
+			obj.data.Logf("graph: "+lang.Name+": "+format, v...)
+		})
+		obj.currentPuppetGraph.Logf(func(format string, v ...interface{}) {
+			obj.data.Logf("graph: "+puppet.Name+": "+format, v...)
+		})
+		if err == nil {
+			g.Logf(func(format string, v ...interface{}) {
+				obj.data.Logf("graph: "+Name+": "+format, v...)
+			})
+		}
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		firstLang := false
+		firstPuppet := false
+
+		for {
+			select {
+			case nextChild := <-nextLang:
+				if nextChild.Err != nil {
+					ch <- nextChild
```

Comment:
deadlock! (same below)

because you can't unblock from here if ch doesn't want to read!

instead, save the err and send it below in one spot at the bottom.
You'll probably also have to do a `(!firstLang || !firstPuppet) && err == nil` below instead of the `!firstLang || !firstPuppet` to pass through.